File: Algorithm/Python/647.palindromic-substrings.py
#
# @lc app=leetcode id=647 lang=python
#
# [647] Palindromic Substrings
#

# @lc code=start
class Solution(object):
    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        result, n = 0, len(s)
        dp = [[False]*n for _ in range(n)]
        for i in range(n-1, -1, -1):
            for j in range(i, n, 1):
                if s[i] == s[j] and (i+1 >= j-1 or dp[i+1][j-1]):
                    dp[i][j] = True
                    result += 1
        return result
        
# @lc code=end

File: Algorithm/C++/647.palindromic-substrings.cpp
/*
 * @lc app=leetcode id=647 lang=cpp
 *
 * [647] Palindromic Substrings
 */

// @lc code=start

#include <string>
#include <vector>

class Solution {
public:
    int countSubstrings(std::string s) {
        int n = s.size(), result = 0;
        std::vector<std::vector<bool>> dp(n, std::vector<bool>(n, false));
        for (int i = n-1; i >= 0; i--) {
            for (int j = i; j < n; j++) {
                if (s[i] == s[j] && (i+1 >= j-1 || dp[i+1][j-1])) {
                    dp[i][j] = true;
                    result += 1;
                }
            }
        }
        return result;
    }
};
// @lc code=end

